sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs

```diff
@@ -12,27 +12,36 @@
 using System.Globalization;
 using System.Linq.Expressions;
 using System.Reflection;
+using System.Threading;
 
 namespace Azure.Core.Pipeline
 {
     internal readonly struct DiagnosticScope : IDisposable
     {
+        private static readonly AsyncLocal<bool> ClientCustomerScopeStarted = new AsyncLocal<bool>()
+        {
+            Value = false
+        };
+
         private static readonly ConcurrentDictionary<string, object?> ActivitySources = new();
 
         private readonly ActivityAdapter? _activityAdapter;
+        private readonly bool _suppressNestedClientScopes;
 
-        internal DiagnosticScope(string ns, string scopeName, DiagnosticListener source, ActivityKind kind)
+        internal DiagnosticScope(string ns, string scopeName, DiagnosticListener source, ActivityKind kind, bool suppressNestedClientScopes) :
+            this(scopeName, source, null, GetActivitySource(ns, scopeName), kind, suppressNestedClientScopes)
         {
-            var activitySource = GetActivitySource(ns, scopeName);
-
-            IsEnabled = source.IsEnabled() || ActivityExtensions.ActivitySourceHasListeners(activitySource);
-
-            _activityAdapter = IsEnabled ? new ActivityAdapter(activitySource, source, scopeName, kind, null) : null;
         }
 
-        internal DiagnosticScope(string scopeName, DiagnosticListener source, object? diagnosticSourceArgs, object? activitySource, ActivityKind kind)
+        internal DiagnosticScope(string scopeName, DiagnosticListener source, object? diagnosticSourceArgs, object? activitySource, ActivityKind kind, bool suppressNestedClientScopes)
         {
             IsEnabled = source.IsEnabled() || ActivityExtensions.ActivitySourceHasListeners(activitySource);
+            _suppressNestedClientScopes = suppressNestedClientScopes && (kind == ActivityKind.Internal || kind == ActivityKind.Client);
+            // ActivityKind.Internal and Client both can represent public API calls.
+            if (_suppressNestedClientScopes)
+            {
+                IsEnabled &= !ClientCustomerScopeStarted.Value;
```

Comment:
So with this, would we be preventing Scope2 from being created? 

- Create Scope1
- Start Scope 1
- Create Scope2
- Dispose Scope 1
- Start Scope 2
- Dispose Scope 2

I'm wondering if this check should be in Start rather than in the constructor.

Reply:
I want to avoid creating ActivityAdapter to save a bit of perf. Also the scope should have `IsEnabled` false so that we don't waste any perf populating attributes on it.

Reply:
Yeah, that's fair - it is definitely an edge case.